include/merlin_hashtable.cuh

```diff
@@ -272,84 +309,97 @@ class HashTable {
+                table_, keys, value_vectors, table_->buckets,
+                table_->buckets_size, table_->bucket_max_size,
                 table_->buckets_num, N);
       } else {
         upsert_kernel_with_io<key_type, vector_type, meta_type, DIM, TILE_SIZE>
             <<<grid_size, block_size, 0, stream>>>(
-                table_, keys, reinterpret_cast<const vector_type*>(values),
-                metas, table_->buckets, table_->buckets_size,
-                table_->bucket_max_size, table_->buckets_num, N);
+                table_, keys, value_vectors, metas, table_->buckets,
+                table_->buckets_size, table_->bucket_max_size,
+                table_->buckets_num, N);
       }
     } else {
-      vector_type** d_dst = nullptr;
-      int* d_src_offset = nullptr;
+      auto device_ws = device_memory_pool_->get_workspace<2>(stream);
+      auto d_dst = device_ws.get<vector_type**>(0);
+      auto d_src_offset = device_ws.get<int*>(1);
 
-      CUDA_CHECK(cudaMallocAsync(&d_dst, n * sizeof(vector_type*), stream));
-      CUDA_CHECK(cudaMemsetAsync(d_dst, 0, n * sizeof(vector_type*), stream));
-      CUDA_CHECK(cudaMallocAsync(&d_src_offset, n * sizeof(int), stream));
-      CUDA_CHECK(cudaMemsetAsync(d_src_offset, 0, n * sizeof(int), stream));
+      HostMemoryPool::StaticWorkspace<3> host_ws;
+      vector_type** h_dst;
+      vector_type* h_values;
+      int* h_src_offset;
 
-      {
-        const size_t block_size = options_.block_size;
-        const size_t N = n * TILE_SIZE;
-        const size_t grid_size = SAFE_GET_GRID_SIZE(N, block_size);
+      size_t max_bs;
+      if (options_.io_by_cpu) {
+        host_ws = std::move(host_memory_pool_->get_workspace<3>(stream));
+        h_dst = host_ws.get<vector_type**>(0);
+        h_values = host_ws.get<vector_type*>(1);
+        h_src_offset = host_ws.get<int*>(2);
 
-        if (metas == nullptr) {
-          upsert_kernel<key_type, vector_type, meta_type, DIM, TILE_SIZE>
-              <<<grid_size, block_size, 0, stream>>>(
-                  table_, keys, d_dst, table_->buckets, table_->buckets_size,
-                  table_->bucket_max_size, table_->buckets_num, d_src_offset,
-                  N);
-        } else {
-          upsert_kernel<key_type, vector_type, meta_type, DIM, TILE_SIZE>
-              <<<grid_size, block_size, 0, stream>>>(
-                  table_, keys, d_dst, metas, table_->buckets,
-                  table_->buckets_size, table_->bucket_max_size,
-                  table_->buckets_num, d_src_offset, N);
-        }
+        max_bs = device_memory_pool_->max_batch_size(
+            std::max(sizeof(vector_type), sizeof(void*)));
+      } else {
+        max_bs = device_memory_pool_->max_batch_size<void*>();
       }
 
-      {
-        thrust::device_ptr<uintptr_t> d_dst_ptr(
-            reinterpret_cast<uintptr_t*>(d_dst));
-        thrust::device_ptr<int> d_src_offset_ptr(d_src_offset);
+      for (size_t i = 0; i < n; i += max_bs) {
```

Comment:
The for loop is strange, and why not transfer the n to the pool before, as you know, the callers almost don't change the batch size when training.

Reply:
We compute the maximum batch size we can do with the `buffer_size` of the pool and then loop over the input data. We reuse the same buffer in each iteration. 

There is no need to transfer that number to the pool. I think you got confused with the frequent use of the variable `n` / `N` (I wanted to change that -- remember? -- but both you and kimi disagreed ;-).

Reply:
The `for` loop is not necessary, will definitely hurt the performance. I think our goal is not to avoid completely memory allocating dynamically just not so frequently. 
`n` has included the requested memory size clearly, we should use it.

Reply:
At least, it would be a better policy to sample the `n` that avoid the for-loop happen, cause the n should be fixed very likely.

Reply:
As said on Slack, the user can specify the a larger `buffer_size` in the pool's options if they have large `n`.

Reply:
The pool needs it to be fixed. That is the essence of the pool. If you want to change it, you have to recreate the pool with new buffers.

If you tune the value online, you only cause a lot of other issues. The user can fine-tune the value for performance if they want. But that is up to the user. The current configuration value is super simple and intuitive. If you put in any automatic size-tuning, you make things super complicated and counter intuitive to the user. I want to give the power to the user to decide what the maximum transaction batch size shall be.

Reply:
I think more, the pool feature is only for internal performance, and it's no need for the end-users to know it and configure it. We'd better not expose the option as a public API. That's confusing the end-users.